examples/deep_learning/flax_resnet.py

```diff
@@ -192,48 +194,51 @@ def loss_fun(params, l2reg, data, aux):
     # batch_stats will be stored in state.aux
     return loss, net_state["batch_stats"]
 
-  def print_evaluation(params, state, l2reg, data, aux):
-    if state.iter_num % 10 == 0:
-      # Periodically evaluate the model on the train and test sets.
-      train_acc, train_loss = accuracy_and_loss(params, l2reg, data, aux)
-      test_acc, test_loss = accuracy_and_loss(params, l2reg, next(test_ds), aux)
-
-      train_acc = jax.device_get(train_acc)
-      train_loss = jax.device_get(train_loss)
-      test_acc = jax.device_get(test_acc)
-      test_loss = jax.device_get(test_loss)
-
-      print(f"[Step {state.iter_num}] "
-            f"Train acc: {train_acc:.3f}, loss: {train_loss:.3f}. "
-            f"Test acc: {test_acc:.3f}, loss: {test_loss:.3f}.")
-
-    return params, state
-
   # Initialize solver.
   opt = optax.sgd(learning_rate=FLAGS.learning_rate,
                   momentum=FLAGS.momentum,
                   nesterov=True)
 
   # We need has_aux=True because loss_fun returns batch_stats.
-  solver = OptaxSolver(opt=opt, fun=loss_fun, maxiter=FLAGS.maxiter,
-                       pre_update=print_evaluation, has_aux=True)
+  solver = OptaxSolver(opt=opt, fun=loss_fun, maxiter=FLAGS.epochs * iter_per_epoch, has_aux=True)
 
   # Initialize parameters.
   rng = jax.random.PRNGKey(0)
   init_vars = net.init(rng, jnp.zeros(input_shape), train=True)
   params = init_vars["params"]
   batch_stats = init_vars["batch_stats"]
+  start = datetime.now().replace(microsecond=0)
 
   # Run training loop.
   state = solver.init_state(params)
 
-  for _ in range(FLAGS.maxiter):
-    params, state = solver.update(params=params,
+  for _ in range(solver.maxiter):
+    train_minibatch = next(train_ds)
+
+    if state.iter_num % iter_per_epoch == iter_per_epoch - 1:
+      # Once per epoch evaluate the model on the train and test sets.
+      train_acc, train_loss = accuracy_and_loss(params, FLAGS.l2reg, train_minibatch, batch_stats)
+      test_acc, test_loss = accuracy_and_loss(params, FLAGS.l2reg, next(test_ds), batch_stats)
+
+      train_acc = jax.device_get(train_acc)
+      train_loss = jax.device_get(train_loss)
+      test_acc = jax.device_get(test_acc)
+      test_loss = jax.device_get(test_loss)
+      # time elapsed without microseconds
+      time_elapsed = (datetime.now().replace(microsecond=0) - start)
+
+      print(f"[Epoch {state.iter_num // (iter_per_epoch+1)}/{FLAGS.epochs}] "
+            f"Train acc: {train_acc:.3f}, train loss: {train_loss:.3f}. "
+            f"Test acc: {test_acc:.3f}, test loss: {test_loss:.3f}. "
+            f"Time elapsed: {time_elapsed}")
+
+    params, state = jax.jit(solver.update)(params=params,
```

Comment:
I hadn't noticed but you're recompiling on every iteration. Also, when/if we implement #117, this will no longer be necessary. CC @fllinares

Reply:
huh, good catch. Performance wise I didn't observe much overhead, with the GPU utilization around 95% . 

Is there a workaround I should use? Perhaps define an ```_update = jax.jit(solver.update)``` function outside of the for loop?

Reply:
Yep exactly (maybe use the name `jitted_update` to be more explicit)